[core] new luautil function: getEntitiesInRange()

MowFord opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Sparked by [this comment](https://github.com/LandSandBoat/server/pull/4868#discussion_r1440575060), I've created a more generic binding for finding entities in range of an entity, using the existing `CTargetFind` class.

I'm putting this in draft mode while I test it, but preliminary testing seems to work (the benefits of using pre-existing systems)

This PR lays some groundwork:
- Creates a missing enum essential to skills/abilities: AOE_TYPE
- Copies all targetfind enums into lua space, with some comments about their caveats

The PR's main change: `getEntitiesInRange()`
- This binding first does basic checks and returns an empty table if anything essential is missing with the parameters
- It then calls a `PAI->Targetfind` flow similar to spells or mobskills
- the return is always a table, empty if no targets were found within the given targetfind parameters

## Steps to test these changes

<!-- Clear and detailed steps to test your changes here -->

MowFord commented:
targetfind is funny about conals in that it always adds the primary target my work in #5745 will make it more reliable if one needed to use this binding for raw conal findings, but in general the binding seems to work well.

It's written explicitly to be used for Dark Ixion, so I'm going to start poking at it there

MowFord commented:
I've done some testing and it seems to do what's on the tin here. Ixion functions with the new binding, but will benefit greatly from the extra `TARGETTYPE` in #5745

zach2good commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -10236,6 +10236,98 @@ void CLuaBaseEntity::recalculateAbilitiesTable()
+sol::table CLuaBaseEntity::getEntitiesInRange(CLuaBaseEntity* PLuaEntityTarget, sol::variadic_args va)
+{
+    auto* baseEntity = dynamic_cast<CBattleEntity*>(m_PBaseEntity);
+    auto  players    = lua.create_table();
```
If this is targetfind-based, it can find more than players depending on the flags, no?

MowFord replied:
Good catch, yes. Started from my old binding. I'll update the variable

MowFord replied:
Done

zach2good commented on src/map/lua/lua_baseentity.h:
```diff
@@ -514,8 +514,9 @@ class CLuaBaseEntity
-    void recalculateAbilitiesTable();
+    void       recalculateSkillsTable();
+    void       recalculateAbilitiesTable();
+    sol::table getEntitiesInRange(CLuaBaseEntity* PLuaEntityTarget, sol::variadic_args va);
```
`auto getEntitiesInRange(CLuaBaseEntity* PLuaEntityTarget, sol::variadic_args va) -> sol::table;`

zach2good commented on src/map/lua/lua_baseentity.cpp:
```diff
@@ -10236,6 +10236,98 @@ void CLuaBaseEntity::recalculateAbilitiesTable()
+ *  Function: getEntitiesInRange()
+ *  Purpose : Returns a Lua table of entities within range of the base entity
+ *  Example : local players = npc:getEntitiesInRange(target, aoeType, radiusOrigin, distance, findFlags, validTargets)
+ *  Notes   : if the passed distance is nil or 0 (as well as many other combinations), empty table will be returned
```
I would add a _big fat note_ here, saying something to the tune of: "APART FROM THIS ONE EXCEPTION, ALL TARGET AND PATH FINDING SHOULD BE DONE EXCLUSIVELY IN CORE"

zach2good replied:
and say that this is used for Ixion, and possibly there's another usage for this in assault? I don't remember

MowFord replied:
Can do

zach2good replied:
And another sub-note: If you have to do expensive operations that are meant for core in Lua, you should be doing them as infrequently as possible: In a mobskill that locks itself off, or anything that isn't on-tick, etc.

MowFord replied:
Updated the description

MowFord committed:
new luautil function: getEntitiesInRange()

get all targets in a sphere around an entity
utilizes targetfind on the entity with parameters for aoe type, validtargets, etc

zach2good commented:
For any passers-by: This is logic added using `TargetFind`. Future optimisations using octrees/spatial hashing will happen INSIDE `TargetFind`.
